// ngx-material-timepicker/esm5/src/app/material-timepicker/pipes/time-parser.pipe.js
import * as tslib_1 from "tslib";
import { Inject, Injectable, Pipe } from '@angular/core';
import { NUMBERING_SYSTEM, TIME_LOCALE } from '../tokens/time-locale.token';
import { TimeUnit } from '../models/time-unit.enum';
import { DateTime } from 'luxon';
var TimeParserPipe = /** @class */ (function () {
    function TimeParserPipe(locale, numberingSystem) {
        this.locale = locale;
        this.numberingSystem = numberingSystem;
    }
    TimeParserPipe.prototype.transform = function (time, timeUnit) {
        if (timeUnit === void 0) { timeUnit = TimeUnit.HOUR; }
        if (time == null || time === '') {
            return '';
        }
        if (!isNaN(+time)) {
            return time;
        }
        if (timeUnit === TimeUnit.MINUTE) {
            return this.parseTime(time, 'm', 'minute');
        }
        return this.parseTime(time, 'H', 'hour');
    };
    TimeParserPipe.prototype.parseTime = function (time, format, timeMeasure) {
        var parsedTime = DateTime.fromFormat(String(time), format, {
            numberingSystem: this.numberingSystem,
            locale: this.locale
        })[timeMeasure];
        if (!isNaN(parsedTime)) {
            return parsedTime;
        }
        throw new Error("Cannot parse time - " + time);
    };
    TimeParserPipe.ctorParameters = function () { return [
        { type: String, decorators: [{ type: Inject, args: [TIME_LOCALE,] }] },
        { type: String, decorators: [{ type: Inject, args: [NUMBERING_SYSTEM,] }] }
    ]; };
    TimeParserPipe = tslib_1.__decorate([
        Pipe({
            name: 'timeParser'
        }),
        Injectable(),
        tslib_1.__param(0, Inject(TIME_LOCALE)), tslib_1.__param(1, Inject(NUMBERING_SYSTEM))
    ], TimeParserPipe);
    return TimeParserPipe;
}());
export { TimeParserPipe };
