// ngx-scrollbar/esm2020/reached-event/ng-scrollbar-reached.module.mjs
import { NgModule } from '@angular/core';
import { BidiModule } from '@angular/cdk/bidi';
import { NgScrollbarModule } from 'ngx-scrollbar';
import { NgScrollbarReachedTop, NgScrollbarReachedBottom, NgScrollbarReachedStart, NgScrollbarReachedEnd } from './ng-scrollbar-reached';
import * as i0 from "@angular/core";
export class NgScrollbarReachedModule {
}
NgScrollbarReachedModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.1.4", ngImport: i0, type: NgScrollbarReachedModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgScrollbarReachedModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.1.4", ngImport: i0, type: NgScrollbarReachedModule, declarations: [NgScrollbarReachedTop,
        NgScrollbarReachedBottom,
        NgScrollbarReachedStart,
        NgScrollbarReachedEnd], imports: [NgScrollbarModule,
        BidiModule], exports: [NgScrollbarReachedTop,
        NgScrollbarReachedBottom,
        NgScrollbarReachedStart,
        NgScrollbarReachedEnd] });
NgScrollbarReachedModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.1.4", ngImport: i0, type: NgScrollbarReachedModule, imports: [NgScrollbarModule,
        BidiModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.1.4", ngImport: i0, type: NgScrollbarReachedModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        NgScrollbarModule,
                        BidiModule
                    ],
                    declarations: [
                        NgScrollbarReachedTop,
                        NgScrollbarReachedBottom,
                        NgScrollbarReachedStart,
                        NgScrollbarReachedEnd
                    ],
                    exports: [
                        NgScrollbarReachedTop,
                        NgScrollbarReachedBottom,
                        NgScrollbarReachedStart,
                        NgScrollbarReachedEnd
                    ]
                }]
        }] });
