// ngx-scrollbar/esm2020/lib/hide-native-scrollbar/hide-native-scrollbar.mjs
import { Directive, RendererStyleFlags2 } from '@angular/core';
import { Subscription } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "./native-scrollbar-size-factory";
export class HideNativeScrollbar {
    constructor(el, renderer, hideNativeScrollbar) {
        this.renderer = renderer;
        this.hideNativeScrollbar = hideNativeScrollbar;
        this._subscriber = Subscription.EMPTY;
        this._subscriber = hideNativeScrollbar.scrollbarSize.subscribe((size) => {
            this.renderer.setStyle(el.nativeElement, '--native-scrollbar-size', `-${size}px`, RendererStyleFlags2.DashCase);
        });
    }
    ngOnDestroy() {
        this._subscriber.unsubscribe();
    }
}
HideNativeScrollbar.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.1.4", ngImport: i0, type: HideNativeScrollbar, deps: [{ token: i0.ElementRef }, { token: i0.Renderer2 }, { token: i1.NativeScrollbarSizeFactory }], target: i0.ɵɵFactoryTarget.Directive });
HideNativeScrollbar.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.1.4", type: HideNativeScrollbar, selector: "[hideNativeScrollbar]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.1.4", ngImport: i0, type: HideNativeScrollbar, decorators: [{
            type: Directive,
            args: [{
                    selector: '[hideNativeScrollbar]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.Renderer2 }, { type: i1.NativeScrollbarSizeFactory }]; } });
